### PR TITLE
[PyCDE] Add `obj_to_value`, which converts Python objects to MLIR Values

### PR DESCRIPTION
**Pull request by teqdruid**
Also provide (and use) a PyCDE-specific `OpOperandConnect` which provide a `connect()` method which uses `obj_to_value` to accept a Python object.

### CODE REVIEW

**Comment by teqdruid** [comment]
@stellaraccident I could use some help debugging an assertion I'm getting on this change, though it seems unrelated to this change. Interesting things to note:
- I only get it in the `polynomial.py` test though I'm also using this change in an internal project.
- The minimal change to make this work vs. break boils down to creating a temporary operation (`TemporaryBackedge`) and then erasing it vs not creating it at all.
- This error goes away when step through code in a python debugger, but not when I run it in the debugger without stepping through some python code. It also goes away when I run python + Valgrind.

Here's the assertion: 

```
python3.9: /home/jodemme/circt/llvm/mlir/lib/Bindings/Python/IRCore.cpp:809: static mlir::python::PyOperationRef mlir::python::PyOperation::createDetached(mlir::python::PyMlirContextRef, MlirOperation, py::object): Assertion `liveOperations.count(operation.ptr) == 0 && "cannot create detached operation that already exists"' failed.
 #0 0x00007f09d0d425da llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/jodemme/circt/llvm/llvm/lib/Support/Unix/Signals.inc:565:11
 #1 0x00007f09d0d427ab PrintStackTraceSignalHandler(void*) /home/jodemme/circt/llvm/llvm/lib/Support/Unix/Signals.inc:632:1
 #2 0x00007f09d0d40d7b llvm::sys::RunSignalHandlers() /home/jodemme/circt/llvm/llvm/lib/Support/Signals.cpp:75:5
 #3 0x00007f09d0d42f21 SignalHandler(int) /home/jodemme/circt/llvm/llvm/lib/Support/Unix/Signals.inc:407:1
 #4 0x00007f09d4be1140 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x14140)
 #5 0x00007f09d48a6ce1 raise ./signal/../sysdeps/unix/sysv/linux/raise.c:51:1
 #6 0x00007f09d4890537 abort ./stdlib/abort.c:81:7
 #7 0x00007f09d489040f get_sysdep_segment_value ./intl/loadmsgcat.c:509:8
 #8 0x00007f09d489040f _nl_load_domain ./intl/loadmsgcat.c:970:34
 #9 0x00007f09d489f662 (/lib/x86_64-linux-gnu/libc.so.6+0x34662)
#10 0x00007f09c4f65409 mlir::python::PyOperation::createDetached(mlir::python::PyObjectRef<mlir::python::PyMlirContext>, MlirOperation, pybind11::object) /home/jodemme/circt/llvm/mlir/lib/Bindings/Python/IRCore.cpp:812:3
#11 0x00007f09c4f679d4 mlir::python::PyOperation::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, llvm::Optional<std::vector<mlir::python::PyType*, std::allocator<mlir::python::PyType*> > >, llvm::Optional<std::vector<mlir::python::PyValue*, std::allocator<mlir::python::PyValue*> > >, llvm::Optional<pybind11::dict>, llvm::Optional<std::vector<mlir::python::PyBlock*, std::allocator<mlir::python::PyBlock*> > >, int, mlir::python::DefaultingPyLocation, pybind11::object) /home/jodemme/circt/llvm/mlir/lib/Bindings/Python/IRCore.cpp:1023:7
#12 0x00007f09c4f6b716 mlir::python::PyOpView::buildGeneric(pybind11::object, pybind11::list, pybind11::list, llvm::Optional<pybind11::dict>, llvm::Optional<std::vector<mlir::python::PyBlock*, std::allocator<mlir::python::PyBlock*> > >, llvm::Optional<int>, mlir::python::DefaultingPyLocation, pybind11::object) /home/jodemme/circt/llvm/mlir/lib/Bindings/Python/IRCore.cpp:1317:10
#13 0x00007f09c4ff150d pybind11::object pybind11::detail::argument_loader<pybind11::object, pybind11::list, pybind11::list, llvm::Optional<pybind11::dict>, llvm::Optional<std::vector<mlir::python::PyBlock*, std::allocator<mlir::python::PyBlock*> > >, llvm::Optional<int>, mlir::python::DefaultingPyLocation, pybind11::object>::call_impl<pybind11::object, pybind11::object (*&)(pybind11::object, pybind11::list, pybind11::list, llvm::Optional<pybind11::dict>, llvm::Optional<std::vector<mlir::python::PyBlock*, std::allocator<mlir::python::PyBlock*> > >, llvm::Optional<int>, mlir::python::DefaultingPyLocation, pybind11::object), 0ul, 1ul, 2ul, 3ul, 4ul, 5ul, 6ul, 7ul, pybind11::detail::void_type>(pybind11::object (*&)(pybind11::object, pybind11::list, pybind11::list, llvm::Optional<pybind11::dict>, llvm::Optional<std::vector<mlir::python::PyBlock*, std::allocator<mlir::python::PyBlock*> > >, llvm::Optional<int>, mlir::python::DefaultingPyLocation, pybind11::object), std::integer_sequence<unsigned long, 0ul, 1ul, 2ul, 3ul, 4ul, 5ul, 6ul, 7ul>, pybind11::detail::void_type&&) && /home/jodemme/.local/lib/python3.9/site-packages/pybind11/include/pybind11/cast.h:2042:16
#14 0x00007f09c4ff03cf std::enable_if<!(std::is_void<pybind11::object>::value), pybind11::object>::type pybind11::detail::argument_loader<pybind11::object, pybind11::list, pybind11::list, llvm::Optional<pybind11::dict>, llvm::Optional<std::vector<mlir::python::PyBlock*, std::allocator<mlir::python::PyBlock*> > >, llvm::Optional<int>, mlir::python::DefaultingPyLocation, pybind11::object>::call<pybind11::object, pybind11::detail::void_type, pybind11::object (*&)(pybind11::object, pybind11::list, pybind11::list, llvm::Optional<pybind11::dict>, llvm::Optional<std::vector<mlir::python::PyBlock*, std::allocator<mlir::python::PyBlock*> > >, llvm::Optional<int>, mlir::python::DefaultingPyLocation, pybind11::object)>(pybind11::object (*&)(pybind11::object, pybind11::list, pybind11::list, llvm::Optional<pybind11::dict>, llvm::Optional<std::vector<mlir::python::PyBlock*, std::allocator<mlir::python::PyBlock*> > >, llvm::Optional<int>, mlir::python::DefaultingPyLocation, pybind11::object)) && /home/jodemme/.local/lib/python3.9/site-packages/pybind11/include/pybind11/cast.h:2014:42
#15 0x00007f09c4ff0189 void pybind11::cpp_function::initialize<pybind11::object (*&)(pybind11::object, pybind11::list, pybind11::list, llvm::Optional<pybind11::dict>, llvm::Optional<std::vector<mlir::python::PyBlock*, std::allocator<mlir::python::PyBlock*> > >, llvm::Optional<int>, mlir::python::DefaultingPyLocation, pybind11::object), pybind11::object, pybind11::object, pybind11::list, pybind11::list, llvm::Optional<pybind11::dict>, llvm::Optional<std::vector<mlir::python::PyBlock*, std::allocator<mlir::python::PyBlock*> > >, llvm::Optional<int>, mlir::python::DefaultingPyLocation, pybind11::object, pybind11::arg, pybind11::arg_v, pybind11::arg_v, pybind11::arg_v, pybind11::arg_v, pybind11::arg_v, pybind11::arg_v, pybind11::arg_v, char const*>(pybind11::object (*&)(pybind11::object, pybind11::list, pybind11::list, llvm::Optional<pybind11::dict>, llvm::Optional<std::vector<mlir::python::PyBlock*, std::allocator<mlir::python::PyBlock*> > >, llvm::Optional<int>, mlir::python::DefaultingPyLocation, pybind11::object), pybind11::object (*)(pybind11::object, pybind11::list, pybind11::list, llvm::Optional<pybind11::dict>, llvm::Optional<std::vector<mlir::python::PyBlock*, std::allocator<mlir::python::PyBlock*> > >, llvm::Optional<int>, mlir::python::DefaultingPyLocation, pybind11::object), pybind11::arg const&, pybind11::arg_v const&, pybind11::arg_v const&, pybind11::arg_v const&, pybind11::arg_v const&, pybind11::arg_v const&, pybind11::arg_v const&, pybind11::arg_v const&, char const* const&)::'lambda'(pybind11::detail::function_call&)::operator()(pybind11::detail::function_call&) const /home/jodemme/.local/lib/python3.9/site-packages/pybind11/include/pybind11/pybind11.h:193:52
#16 0x00007f09c4ff0075 void pybind11::cpp_function::initialize<pybind11::object (*&)(pybind11::object, pybind11::list, pybind11::list, llvm::Optional<pybind11::dict>, llvm::Optional<std::vector<mlir::python::PyBlock*, std::allocator<mlir::python::PyBlock*> > >, llvm::Optional<int>, mlir::python::DefaultingPyLocation, pybind11::object), pybind11::object, pybind11::object, pybind11::list, pybind11::list, llvm::Optional<pybind11::dict>, llvm::Optional<std::vector<mlir::python::PyBlock*, std::allocator<mlir::python::PyBlock*> > >, llvm::Optional<int>, mlir::python::DefaultingPyLocation, pybind11::object, pybind11::arg, pybind11::arg_v, pybind11::arg_v, pybind11::arg_v, pybind11::arg_v, pybind11::arg_v, pybind11::arg_v, pybind11::arg_v, char const*>(pybind11::object (*&)(pybind11::object, pybind11::list, pybind11::list, llvm::Optional<pybind11::dict>, llvm::Optional<std::vector<mlir::python::PyBlock*, std::allocator<mlir::python::PyBlock*> > >, llvm::Optional<int>, mlir::python::DefaultingPyLocation, pybind11::object), pybind11::object (*)(pybind11::object, pybind11::list, pybind11::list, llvm::Optional<pybind11::dict>, llvm::Optional<std::vector<mlir::python::PyBlock*, std::allocator<mlir::python::PyBlock*> > >, llvm::Optional<int>, mlir::python::DefaultingPyLocation, pybind11::object), pybind11::arg const&, pybind11::arg_v const&, pybind11::arg_v const&, pybind11::arg_v const&, pybind11::arg_v const&, pybind11::arg_v const&, pybind11::arg_v const&, pybind11::arg_v const&, char const* const&)::'lambda'(pybind11::detail::function_call&)::__invoke(pybind11::detail::function_call&) /home/jodemme/.local/lib/python3.9/site-packages/pybind11/include/pybind11/pybind11.h:170:21
#17 0x00007f09c4e9311c pybind11::cpp_function::dispatcher(_object*, _object*, _object*) /home/jodemme/.local/lib/python3.9/site-packages/pybind11/include/pybind11/pybind11.h:767:30
#18 0x000000000053f350 (/usr/bin/python3.9+0x53f350)
#19 0x000000000051d89b _PyObject_MakeTpCall (/usr/bin/python3.9+0x51d89b)
#20 0x000000000053bf25 (/usr/bin/python3.9+0x53bf25)
#21 0x0000000000513456 _PyEval_EvalFrameDefault (/usr/bin/python3.9+0x513456)
#22 0x00000000005106ed (/usr/bin/python3.9+0x5106ed)
#23 0x0000000000528d21 _PyFunction_Vectorcall (/usr/bin/python3.9+0x528d21)
#24 0x00000000005377ce (/usr/bin/python3.9+0x5377ce)
#25 0x000000000051dddf (/usr/bin/python3.9+0x51dddf)
#26 0x00007f09d4371e33 pybind11_meta_call /home/jodemme/.local/lib/python3.9/site-packages/pybind11/include/pybind11/detail/class.h:175:15
#27 0x000000000053c3c9 PyObject_Call (/usr/bin/python3.9+0x53c3c9)
#28 0x0000000000513e8b _PyEval_EvalFrameDefault (/usr/bin/python3.9+0x513e8b)
#29 0x00000000005106ed (/usr/bin/python3.9+0x5106ed)
#30 0x0000000000528d21 _PyFunction_Vectorcall (/usr/bin/python3.9+0x528d21)
#31 0x000000000053bcfb (/usr/bin/python3.9+0x53bcfb)
#32 0x0000000000513456 _PyEval_EvalFrameDefault (/usr/bin/python3.9+0x513456)
#33 0x00000000005106ed (/usr/bin/python3.9+0x5106ed)
#34 0x0000000000528d21 _PyFunction_Vectorcall (/usr/bin/python3.9+0x528d21)
#35 0x00000000005377ce (/usr/bin/python3.9+0x5377ce)
#36 0x000000000051d6f5 _PyObject_MakeTpCall (/usr/bin/python3.9+0x51d6f5)
#37 0x0000000000517a0f _PyEval_EvalFrameDefault (/usr/bin/python3.9+0x517a0f)
#38 0x00000000005106ed (/usr/bin/python3.9+0x5106ed)
#39 0x0000000000528d21 _PyFunction_Vectorcall (/usr/bin/python3.9+0x528d21)
#40 0x000000000053bcfb (/usr/bin/python3.9+0x53bcfb)
#41 0x000000000053c361 PyObject_Call (/usr/bin/python3.9+0x53c361)
#42 0x0000000000513e8b _PyEval_EvalFrameDefault (/usr/bin/python3.9+0x513e8b)
#43 0x0000000000510fe7 (/usr/bin/python3.9+0x510fe7)
#44 0x0000000000528d21 _PyFunction_Vectorcall (/usr/bin/python3.9+0x528d21)
#45 0x000000000051cf95 _PyObject_FastCallDictTstate (/usr/bin/python3.9+0x51cf95)
#46 0x0000000000538545 _PyObject_Call_Prepend (/usr/bin/python3.9+0x538545)
#47 0x00000000005b81f3 (/usr/bin/python3.9+0x5b81f3)
#48 0x00000000005ce922 PyObject_CallObject (/usr/bin/python3.9+0x5ce922)
#49 0x00007f09d436ca42 pybind11::detail::simple_collector<(pybind11::return_value_policy)1>::call(_object*) const /home/jodemme/.local/lib/python3.9/site-packages/pybind11/include/pybind11/cast.h:2064:19
#50 0x00007f09d43bbcb4 pybind11::object pybind11::detail::object_api<pybind11::handle>::operator()<(pybind11::return_value_policy)1, MlirOperation&>(MlirOperation&) const /home/jodemme/.local/lib/python3.9/site-packages/pybind11/include/pybind11/cast.h:2229:75
#51 0x00007f09d43b98aa callPyFunc(MlirOperation, void*) /home/jodemme/circt/lib/Bindings/Python/MSFTModule.cpp:47:22
#52 0x00007f09d422396e mlirMSFTRegisterGenerator::$_0::operator()(mlir::Operation*) const /home/jodemme/circt/lib/CAPI/Dialect/MSFT.cpp:35:23
#53 0x00007f09d4223912 mlir::Operation* std::__invoke_impl<mlir::Operation*, mlirMSFTRegisterGenerator::$_0&, mlir::Operation*>(std::__invoke_other, mlirMSFTRegisterGenerator::$_0&, mlir::Operation*&&) /bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:60:7
#54 0x00007f09d42238a2 std::enable_if<__and_<std::__not_<std::is_void<mlir::Operation*> >, std::is_convertible<std::__invoke_result<mlirMSFTRegisterGenerator::$_0&, mlir::Operation*>::type, mlir::Operation*> >::value, mlir::Operation*>::type std::__invoke_r<mlir::Operation*, mlirMSFTRegisterGenerator::$_0&, mlir::Operation*>(mlirMSFTRegisterGenerator::$_0&, mlir::Operation*&&) /bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:141:7
#55 0x00007f09d42237b2 std::_Function_handler<mlir::Operation* (mlir::Operation*), mlirMSFTRegisterGenerator::$_0>::_M_invoke(std::_Any_data const&, mlir::Operation*&&) /bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/std_function.h:291:2
#56 0x00007f09d3ed885b std::function<mlir::Operation* (mlir::Operation*)>::operator()(mlir::Operation*) const /bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/std_function.h:622:7
#57 0x00007f09d3eccc3c (anonymous namespace)::OpGenerator::runOnOperation(mlir::Operation*, llvm::SmallSet<llvm::StringRef, 8u, std::less<llvm::StringRef> >) /home/jodemme/circt/lib/Dialect/MSFT/MSFTGenerator.cpp:75:14
#58 0x00007f09d3ed81cb circt::msft::detail::Generators::runOnOperation(mlir::Operation*, llvm::SmallSet<llvm::StringRef, 8u, std::less<llvm::StringRef> >) /home/jodemme/circt/lib/Dialect/MSFT/MSFTGenerator.cpp:94:32
#59 0x00007f09d3ecc694 (anonymous namespace)::RunGeneratorsPass::runOnOperation()::$_0::operator()(mlir::Operation*) const /home/jodemme/circt/lib/Dialect/MSFT/MSFTGenerator.cpp:134:34
#60 0x00007f09d3ecc600 void llvm::function_ref<void (mlir::Operation*)>::callback_fn<(anonymous namespace)::RunGeneratorsPass::runOnOperation()::$_0>(long, mlir::Operation*) /home/jodemme/circt/llvm/llvm/include/llvm/ADT/STLExtras.h:177:5
#61 0x00007f09d11f416c llvm::function_ref<void (mlir::Operation*)>::operator()(mlir::Operation*) const /home/jodemme/circt/llvm/llvm/include/llvm/ADT/STLExtras.h:200:5
#62 0x00007f09d11f3746 mlir::detail::walk(mlir::Operation*, llvm::function_ref<void (mlir::Operation*)>, mlir::WalkOrder) /home/jodemme/circt/llvm/mlir/lib/IR/Visitors.cpp:68:1
#63 0x00007f09d11f36f3 mlir::detail::walk(mlir::Operation*, llvm::function_ref<void (mlir::Operation*)>, mlir::WalkOrder) /home/jodemme/circt/llvm/mlir/lib/IR/Visitors.cpp:61:27
#64 0x00007f09d11f36f3 mlir::detail::walk(mlir::Operation*, llvm::function_ref<void (mlir::Operation*)>, mlir::WalkOrder) /home/jodemme/circt/llvm/mlir/lib/IR/Visitors.cpp:61:27
#65 0x00007f09d3ecc585 std::enable_if<llvm::is_one_of<mlir::Operation*, mlir::Operation*, mlir::Region*, mlir::Block*>::value, void>::type mlir::detail::walk<(mlir::WalkOrder)1, (anonymous namespace)::RunGeneratorsPass::runOnOperation()::$_0, mlir::Operation*, void>(mlir::Operation*, (anonymous namespace)::RunGeneratorsPass::runOnOperation()::$_0&&) /home/jodemme/circt/llvm/mlir/include/mlir/IR/Visitors.h:134:3
#66 0x00007f09d3ecc500 void mlir::Operation::walk<(mlir::WalkOrder)1, (anonymous namespace)::RunGeneratorsPass::runOnOperation()::$_0, void>((anonymous namespace)::RunGeneratorsPass::runOnOperation()::$_0&&) /home/jodemme/circt/llvm/mlir/include/mlir/IR/Operation.h:522:5
#67 0x00007f09d3ecc388 (anonymous namespace)::RunGeneratorsPass::runOnOperation() /home/jodemme/circt/lib/Dialect/MSFT/MSFTGenerator.cpp:133:3
#68 0x00007f09d2aa4b78 mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) /home/jodemme/circt/llvm/mlir/lib/Pass/Pass.cpp:386:21
#69 0x00007f09d2aa5165 mlir::detail::OpToOpPassAdaptor::runPipeline(llvm::iterator_range<llvm::pointee_iterator<std::unique_ptr<mlir::Pass, std::default_delete<mlir::Pass> >*, mlir::Pass> >, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) /home/jodemme/circt/llvm/mlir/lib/Pass/Pass.cpp:445:16
#70 0x00007f09d2aa685e mlir::PassManager::runPasses(mlir::Operation*, mlir::AnalysisManager) /home/jodemme/circt/llvm/mlir/lib/Pass/Pass.cpp:690:10
#71 0x00007f09d2aa6748 mlir::PassManager::run(mlir::Operation*) /home/jodemme/circt/llvm/mlir/lib/Pass/Pass.cpp:670:60
#72 0x00007f09d42f71a2 mlirPassManagerRun /home/jodemme/circt/llvm/mlir/lib/CAPI/IR/Pass.cpp:38:36
#73 0x00007f09c5052bee mlir::python::populatePassManagerSubmodule(pybind11::module_&)::$_4::operator()((anonymous namespace)::PyPassManager&, mlir::python::PyModule&) const /home/jodemme/circt/llvm/mlir/lib/Bindings/Python/Pass.cpp:104:17
#74 0x00007f09c5052b8f void pybind11::detail::argument_loader<(anonymous namespace)::PyPassManager&, mlir::python::PyModule&>::call_impl<void, mlir::python::populatePassManagerSubmodule(pybind11::module_&)::$_4&, 0ul, 1ul, pybind11::detail::void_type>(mlir::python::populatePassManagerSubmodule(pybind11::module_&)::$_4&, std::integer_sequence<unsigned long, 0ul, 1ul>, pybind11::detail::void_type&&) && /home/jodemme/.local/lib/python3.9/site-packages/pybind11/include/pybind11/cast.h:2042:9
#75 0x00007f09c5052886 std::enable_if<std::is_void<void>::value, pybind11::detail::void_type>::type pybind11::detail::argument_loader<(anonymous namespace)::PyPassManager&, mlir::python::PyModule&>::call<void, pybind11::detail::void_type, mlir::python::populatePassManagerSubmodule(pybind11::module_&)::$_4&>(mlir::python::populatePassManagerSubmodule(pybind11::module_&)::$_4&) && /home/jodemme/.local/lib/python3.9/site-packages/pybind11/include/pybind11/cast.h:2020:9
#76 0x00007f09c505279e void pybind11::cpp_function::initialize<mlir::python::populatePassManagerSubmodule(pybind11::module_&)::$_4, void, (anonymous namespace)::PyPassManager&, mlir::python::PyModule&, pybind11::name, pybind11::is_method, pybind11::sibling, char [78]>(mlir::python::populatePassManagerSubmodule(pybind11::module_&)::$_4&&, void (*)((anonymous namespace)::PyPassManager&, mlir::python::PyModule&), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&, char const (&) [78])::'lambda'(pybind11::detail::function_call&)::operator()(pybind11::detail::function_call&) const /home/jodemme/.local/lib/python3.9/site-packages/pybind11/include/pybind11/pybind11.h:193:81
#77 0x00007f09c50526f5 void pybind11::cpp_function::initialize<mlir::python::populatePassManagerSubmodule(pybind11::module_&)::$_4, void, (anonymous namespace)::PyPassManager&, mlir::python::PyModule&, pybind11::name, pybind11::is_method, pybind11::sibling, char [78]>(mlir::python::populatePassManagerSubmodule(pybind11::module_&)::$_4&&, void (*)((anonymous namespace)::PyPassManager&, mlir::python::PyModule&), pybind11::name const&, pybind11::is_method const&, pybind11::sibling const&, char const (&) [78])::'lambda'(pybind11::detail::function_call&)::__invoke(pybind11::detail::function_call&) /home/jodemme/.local/lib/python3.9/site-packages/pybind11/include/pybind11/pybind11.h:170:21
#78 0x00007f09c4e9311c pybind11::cpp_function::dispatcher(_object*, _object*, _object*) /home/jodemme/.local/lib/python3.9/site-packages/pybind11/include/pybind11/pybind11.h:767:30
#79 0x000000000053f350 (/usr/bin/python3.9+0x53f350)
#80 0x000000000051d89b _PyObject_MakeTpCall (/usr/bin/python3.9+0x51d89b)
#81 0x000000000053bf25 (/usr/bin/python3.9+0x53bf25)
#82 0x0000000000516e76 _PyEval_EvalFrameDefault (/usr/bin/python3.9+0x516e76)
#83 0x00000000005106ed (/usr/bin/python3.9+0x5106ed)
#84 0x0000000000528d21 _PyFunction_Vectorcall (/usr/bin/python3.9+0x528d21)
#85 0x0000000000512192 _PyEval_EvalFrameDefault (/usr/bin/python3.9+0x512192)
#86 0x00000000005106ed (/usr/bin/python3.9+0x5106ed)
#87 0x0000000000510497 _PyEval_EvalCodeWithName (/usr/bin/python3.9+0x510497)
#88 0x00000000005f5be3 PyEval_EvalCode (/usr/bin/python3.9+0x5f5be3)
#89 0x0000000000619de7 (/usr/bin/python3.9+0x619de7)
#90 0x0000000000615610 (/usr/bin/python3.9+0x615610)
#91 0x0000000000619d79 (/usr/bin/python3.9+0x619d79)
#92 0x0000000000619816 PyRun_SimpleFileExFlags (/usr/bin/python3.9+0x619816)
#93 0x000000000060d4e3 Py_RunMain (/usr/bin/python3.9+0x60d4e3)
#94 0x00000000005ea6e9 Py_BytesMain (/usr/bin/python3.9+0x5ea6e9)
#95 0x00007f09d4891d0a __libc_start_main ./csu/../csu/libc-start.c:308:16
#96 0x00000000005ea5ea _start (/usr/bin/python3.9+0x5ea5ea)
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Aborted
```

Based on the behaviors listed above, the assertion message, and the stack frame above where it's coming from (`PyOperation::create @ IRCore.cpp:1023`), I think MLIR (via the C API) is returning a pointer to a new operation which the python bindings think is a pointer to an operation which is still alive. What I'm struggling to figure out is why.  Can you provide some tips for me to better focus my debugging?

One additional thing to note is that I'm getting this in the middle of a python callback from a C++ pass which is modifying the IR. (The C++ pass is querying Python asking, "what operation should I substitute for this one?") Both the pass and Python are modifying the IR, but not the same block or region. Could that change the MLIR C API operation pointers, invalidating the `liveOperations` map?

**Comment by mikeurbach** [comment]
@teqdruid I believe Stella's previous comments on the revision where I added the Python `erase` method to Operations might shed some light: https://reviews.llvm.org/D101422. I was having some issues with the `liveOperations` map back then. Also, check out the explicit TODO we decided to leave. Could we be hitting that scenario?

From my debugging on that issue, I don't think something will change the MLIR C API operation pointers, but weird things can happen as pointers get re-used. Consider a case like this: operation created at address X, operation deleted from address X (but address X still in live operations map), operation created with address X re-used by malloc or whatever else is under the hood, assertion because address X still in live operation map. The assertion I dealt with in the above revision was different than the one you are hitting (Python destructor vs. createDetached), but the underlying cause sounds similar.

**Comment by teqdruid** [comment]
> Also, check out the explicit TODO we decided to leave. Could we be hitting that scenario?

Saw that. I don't think so since since we're not erasing any ops with regions.

**Comment by teqdruid** [comment]
> From my debugging on that issue, I don't think something will change the MLIR C API operation pointers, but weird things can happen as pointers get re-used. Consider a case like this: operation created at address X, operation deleted from address X (but address X still in live operations map), operation created with address X re-used by malloc or whatever else is under the hood, assertion because address X still in live operation map

Yeah, this is exactly what I think is going on but I'm at a loss for why. I'll look over the comments in your patch.

**Comment by mikeurbach** [comment]
Indeed the issue is not related to the change in this PR. The demonstration, explanation, and fix is over here: https://github.com/llvm/circt/pull/1439.

**Comment by teqdruid** [comment]
@mikeurbach  With your PR, this no longer fails. I think it's good to go.

File: frontends/PyCDE/src/pycde/module.py [hunk 1]
```diff
@@ -292,11 +295,11 @@ def outputs() -> list[(str, mlir.ir.Type)]:
   # subclasses. Add the names to "don't touch" since they can't be touched
   # (since they implictly call an OpView property) when the attributes are being
   # scanned in the `mod` constructor.
-  for (idx, (name, _)) in enumerate(mod._input_ports):
+  for (idx, (name, type)) in enumerate(mod._input_ports):
     setattr(
         mod, name,
-        property(lambda self, idx=idx: OpOperand(self, idx, self.operands[idx],
-                                                 self)))
+        property(lambda self, type=type, idx=idx:
```
Comment (mikeurbach) [inline frontends/PyCDE/src/pycde/module.py:301]: Is `type` being used anywhere? I believe `OpOperandConnect` is just going to take the type from the `Value` passed as the third argument to the constructor.

**Review by mikeurbach — APPROVED** [review]
Looks good, but I only see tests for the `IntegerType` case in `obj_to_value`. Maybe add some tests for the aggregate and alias use cases? These could potentially be more like unit tests where you just make an `OpOperandConnect` directly and try `connect`ing various thing.

File: frontends/PyCDE/src/pycde/module.py (hunk 1 above)
Comment (teqdruid) [reply frontends/PyCDE/src/pycde/module.py]: Development cruft. Good catch.

**Comment by teqdruid** [comment]
Thanks for the review @mikeurbach !

**Pull request merged**